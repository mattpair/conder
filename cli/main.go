package main

import (
	// "./cmd"
	"fmt"
	"io/ioutil"
	"net/http"
)

func main() {
	out, err := http.Get("https://localhost:7344")
	if err != nil {
		fmt.Println("error", err)
		return
	}
	defer out.Body.Close()
	if out != nil {
		fmt.Println(out)
	}
	body, err := ioutil.ReadAll(out.Body)

	if err != nil {
		fmt.Println("error", err)
	}
	if body != nil {
		fmt.Println(out)
	}

	// cmd.Execute()
}
