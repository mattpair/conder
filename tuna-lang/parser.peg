
initializeConst := 'const' ws+ name=name ws+ equals ws+ value=literal
name := name='[a-zA-Z]+\w*'
equals := '='
ws := '\s'
obj := '\{' ws* '\}'
literal := obj