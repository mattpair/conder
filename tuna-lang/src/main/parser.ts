/* AutoGenerated Code, changes may be overwritten
* INPUT GRAMMAR:
* globals := {value={constant | func} {ws+ | '$'}}*
* constant := ws* 'const' ws+ name=name ws* equals ws* value=literal
* name := name='[a-zA-Z]+\w*'
* equals := '='
* ws := '\s'
* obj := '\{' ws* fields=fields ws* '\}'
* literal := obj
* newLineOrComma := '\n|,'
* field := name=name ws* ':' value=literal newLineOrComma?
* fields := value=field*
* space := ' '
* executable := ws* value={value={ret} ws+ }*
* ret := 'return'
* func := ws* 'public' space+ 'function' space+ name=name space* args=args ws* '\{' body=executable '\}'
* args := '\(' ws* leadingArgs={name=name newLineOrComma ws*}* ws* lastArg={name}? ws* '\)'
*/
type Nullable<T> = T | null;
type $$RuleType<T> = (log?: (msg: string) => void) => Nullable<T>;
export interface ContextRecorder {
    record(pos: PosInfo, depth: number, result: any, negating: boolean, extraInfo: string[]): void;
}
interface ASTNodeIntf {
    kind: ASTKinds;
}
export enum ASTKinds {
    globals = "globals",
    globals_$0 = "globals_$0",
    globals_$0_$0_1 = "globals_$0_$0_1",
    globals_$0_$0_2 = "globals_$0_$0_2",
    globals_$0_$1_1 = "globals_$0_$1_1",
    globals_$0_$1_2 = "globals_$0_$1_2",
    constant = "constant",
    name = "name",
    equals = "equals",
    ws = "ws",
    obj = "obj",
    literal = "literal",
    newLineOrComma = "newLineOrComma",
    field = "field",
    fields = "fields",
    space = "space",
    executable = "executable",
    executable_$0 = "executable_$0",
    executable_$0_$0 = "executable_$0_$0",
    ret = "ret",
    func = "func",
    args = "args",
    args_$0 = "args_$0",
    args_$1 = "args_$1",
}
export type globals = globals_$0[];
export interface globals_$0 {
    kind: ASTKinds.globals_$0;
    value: globals_$0_$0;
}
export type globals_$0_$0 = globals_$0_$0_1 | globals_$0_$0_2;
export type globals_$0_$0_1 = constant;
export type globals_$0_$0_2 = func;
export type globals_$0_$1 = globals_$0_$1_1 | globals_$0_$1_2;
export type globals_$0_$1_1 = ws[];
export type globals_$0_$1_2 = string;
export interface constant {
    kind: ASTKinds.constant;
    name: name;
    value: literal;
}
export interface name {
    kind: ASTKinds.name;
    name: string;
}
export type equals = string;
export type ws = string;
export interface obj {
    kind: ASTKinds.obj;
    fields: fields;
}
export type literal = obj;
export type newLineOrComma = string;
export interface field {
    kind: ASTKinds.field;
    name: name;
    value: literal;
}
export interface fields {
    kind: ASTKinds.fields;
    value: field[];
}
export type space = string;
export interface executable {
    kind: ASTKinds.executable;
    value: executable_$0[];
}
export interface executable_$0 {
    kind: ASTKinds.executable_$0;
    value: executable_$0_$0;
}
export type executable_$0_$0 = ret;
export type ret = string;
export interface func {
    kind: ASTKinds.func;
    name: name;
    args: args;
    body: executable;
}
export interface args {
    kind: ASTKinds.args;
    leadingArgs: args_$0[];
    lastArg: Nullable<args_$1>;
}
export interface args_$0 {
    kind: ASTKinds.args_$0;
    name: name;
}
export type args_$1 = name;
export class Parser {
    private readonly input: string;
    private pos: PosInfo;
    private negating: boolean = false;
    constructor(input: string) {
        this.pos = {overallPos: 0, line: 1, offset: 0};
        this.input = input;
    }
    public reset(pos: PosInfo) {
        this.pos = pos;
    }
    public finished(): boolean {
        return this.pos.overallPos === this.input.length;
    }
    public matchglobals($$dpth: number, $$cr?: ContextRecorder): Nullable<globals> {
        return this.loop<globals_$0>(() => this.matchglobals_$0($$dpth + 1, $$cr), true);
    }
    public matchglobals_$0($$dpth: number, $$cr?: ContextRecorder): Nullable<globals_$0> {
        return this.runner<globals_$0>($$dpth,
            (log) => {
                if (log) {
                    log("globals_$0");
                }
                let $scope$value: Nullable<globals_$0_$0>;
                let $$res: Nullable<globals_$0> = null;
                if (true
                    && ($scope$value = this.matchglobals_$0_$0($$dpth + 1, $$cr)) !== null
                    && this.matchglobals_$0_$1($$dpth + 1, $$cr) !== null
                ) {
                    $$res = {kind: ASTKinds.globals_$0, value: $scope$value};
                }
                return $$res;
            }, $$cr)();
    }
    public matchglobals_$0_$0($$dpth: number, $$cr?: ContextRecorder): Nullable<globals_$0_$0> {
        return this.choice<globals_$0_$0>([
            () => this.matchglobals_$0_$0_1($$dpth + 1, $$cr),
            () => this.matchglobals_$0_$0_2($$dpth + 1, $$cr),
        ]);
    }
    public matchglobals_$0_$0_1($$dpth: number, $$cr?: ContextRecorder): Nullable<globals_$0_$0_1> {
        return this.matchconstant($$dpth + 1, $$cr);
    }
    public matchglobals_$0_$0_2($$dpth: number, $$cr?: ContextRecorder): Nullable<globals_$0_$0_2> {
        return this.matchfunc($$dpth + 1, $$cr);
    }
    public matchglobals_$0_$1($$dpth: number, $$cr?: ContextRecorder): Nullable<globals_$0_$1> {
        return this.choice<globals_$0_$1>([
            () => this.matchglobals_$0_$1_1($$dpth + 1, $$cr),
            () => this.matchglobals_$0_$1_2($$dpth + 1, $$cr),
        ]);
    }
    public matchglobals_$0_$1_1($$dpth: number, $$cr?: ContextRecorder): Nullable<globals_$0_$1_1> {
        return this.loop<ws>(() => this.matchws($$dpth + 1, $$cr), false);
    }
    public matchglobals_$0_$1_2($$dpth: number, $$cr?: ContextRecorder): Nullable<globals_$0_$1_2> {
        return this.regexAccept(String.raw`(?:$)`, $$dpth + 1, $$cr);
    }
    public matchconstant($$dpth: number, $$cr?: ContextRecorder): Nullable<constant> {
        return this.runner<constant>($$dpth,
            (log) => {
                if (log) {
                    log("constant");
                }
                let $scope$name: Nullable<name>;
                let $scope$value: Nullable<literal>;
                let $$res: Nullable<constant> = null;
                if (true
                    && this.loop<ws>(() => this.matchws($$dpth + 1, $$cr), true) !== null
                    && this.regexAccept(String.raw`(?:const)`, $$dpth + 1, $$cr) !== null
                    && this.loop<ws>(() => this.matchws($$dpth + 1, $$cr), false) !== null
                    && ($scope$name = this.matchname($$dpth + 1, $$cr)) !== null
                    && this.loop<ws>(() => this.matchws($$dpth + 1, $$cr), true) !== null
                    && this.matchequals($$dpth + 1, $$cr) !== null
                    && this.loop<ws>(() => this.matchws($$dpth + 1, $$cr), true) !== null
                    && ($scope$value = this.matchliteral($$dpth + 1, $$cr)) !== null
                ) {
                    $$res = {kind: ASTKinds.constant, name: $scope$name, value: $scope$value};
                }
                return $$res;
            }, $$cr)();
    }
    public matchname($$dpth: number, $$cr?: ContextRecorder): Nullable<name> {
        return this.runner<name>($$dpth,
            (log) => {
                if (log) {
                    log("name");
                }
                let $scope$name: Nullable<string>;
                let $$res: Nullable<name> = null;
                if (true
                    && ($scope$name = this.regexAccept(String.raw`(?:[a-zA-Z]+\w*)`, $$dpth + 1, $$cr)) !== null
                ) {
                    $$res = {kind: ASTKinds.name, name: $scope$name};
                }
                return $$res;
            }, $$cr)();
    }
    public matchequals($$dpth: number, $$cr?: ContextRecorder): Nullable<equals> {
        return this.regexAccept(String.raw`(?:=)`, $$dpth + 1, $$cr);
    }
    public matchws($$dpth: number, $$cr?: ContextRecorder): Nullable<ws> {
        return this.regexAccept(String.raw`(?:\s)`, $$dpth + 1, $$cr);
    }
    public matchobj($$dpth: number, $$cr?: ContextRecorder): Nullable<obj> {
        return this.runner<obj>($$dpth,
            (log) => {
                if (log) {
                    log("obj");
                }
                let $scope$fields: Nullable<fields>;
                let $$res: Nullable<obj> = null;
                if (true
                    && this.regexAccept(String.raw`(?:\{)`, $$dpth + 1, $$cr) !== null
                    && this.loop<ws>(() => this.matchws($$dpth + 1, $$cr), true) !== null
                    && ($scope$fields = this.matchfields($$dpth + 1, $$cr)) !== null
                    && this.loop<ws>(() => this.matchws($$dpth + 1, $$cr), true) !== null
                    && this.regexAccept(String.raw`(?:\})`, $$dpth + 1, $$cr) !== null
                ) {
                    $$res = {kind: ASTKinds.obj, fields: $scope$fields};
                }
                return $$res;
            }, $$cr)();
    }
    public matchliteral($$dpth: number, $$cr?: ContextRecorder): Nullable<literal> {
        return this.matchobj($$dpth + 1, $$cr);
    }
    public matchnewLineOrComma($$dpth: number, $$cr?: ContextRecorder): Nullable<newLineOrComma> {
        return this.regexAccept(String.raw`(?:\n|,)`, $$dpth + 1, $$cr);
    }
    public matchfield($$dpth: number, $$cr?: ContextRecorder): Nullable<field> {
        return this.runner<field>($$dpth,
            (log) => {
                if (log) {
                    log("field");
                }
                let $scope$name: Nullable<name>;
                let $scope$value: Nullable<literal>;
                let $$res: Nullable<field> = null;
                if (true
                    && ($scope$name = this.matchname($$dpth + 1, $$cr)) !== null
                    && this.loop<ws>(() => this.matchws($$dpth + 1, $$cr), true) !== null
                    && this.regexAccept(String.raw`(?::)`, $$dpth + 1, $$cr) !== null
                    && ($scope$value = this.matchliteral($$dpth + 1, $$cr)) !== null
                    && ((this.matchnewLineOrComma($$dpth + 1, $$cr)) || true)
                ) {
                    $$res = {kind: ASTKinds.field, name: $scope$name, value: $scope$value};
                }
                return $$res;
            }, $$cr)();
    }
    public matchfields($$dpth: number, $$cr?: ContextRecorder): Nullable<fields> {
        return this.runner<fields>($$dpth,
            (log) => {
                if (log) {
                    log("fields");
                }
                let $scope$value: Nullable<field[]>;
                let $$res: Nullable<fields> = null;
                if (true
                    && ($scope$value = this.loop<field>(() => this.matchfield($$dpth + 1, $$cr), true)) !== null
                ) {
                    $$res = {kind: ASTKinds.fields, value: $scope$value};
                }
                return $$res;
            }, $$cr)();
    }
    public matchspace($$dpth: number, $$cr?: ContextRecorder): Nullable<space> {
        return this.regexAccept(String.raw`(?: )`, $$dpth + 1, $$cr);
    }
    public matchexecutable($$dpth: number, $$cr?: ContextRecorder): Nullable<executable> {
        return this.runner<executable>($$dpth,
            (log) => {
                if (log) {
                    log("executable");
                }
                let $scope$value: Nullable<executable_$0[]>;
                let $$res: Nullable<executable> = null;
                if (true
                    && this.loop<ws>(() => this.matchws($$dpth + 1, $$cr), true) !== null
                    && ($scope$value = this.loop<executable_$0>(() => this.matchexecutable_$0($$dpth + 1, $$cr), true)) !== null
                ) {
                    $$res = {kind: ASTKinds.executable, value: $scope$value};
                }
                return $$res;
            }, $$cr)();
    }
    public matchexecutable_$0($$dpth: number, $$cr?: ContextRecorder): Nullable<executable_$0> {
        return this.runner<executable_$0>($$dpth,
            (log) => {
                if (log) {
                    log("executable_$0");
                }
                let $scope$value: Nullable<executable_$0_$0>;
                let $$res: Nullable<executable_$0> = null;
                if (true
                    && ($scope$value = this.matchexecutable_$0_$0($$dpth + 1, $$cr)) !== null
                    && this.loop<ws>(() => this.matchws($$dpth + 1, $$cr), false) !== null
                ) {
                    $$res = {kind: ASTKinds.executable_$0, value: $scope$value};
                }
                return $$res;
            }, $$cr)();
    }
    public matchexecutable_$0_$0($$dpth: number, $$cr?: ContextRecorder): Nullable<executable_$0_$0> {
        return this.matchret($$dpth + 1, $$cr);
    }
    public matchret($$dpth: number, $$cr?: ContextRecorder): Nullable<ret> {
        return this.regexAccept(String.raw`(?:return)`, $$dpth + 1, $$cr);
    }
    public matchfunc($$dpth: number, $$cr?: ContextRecorder): Nullable<func> {
        return this.runner<func>($$dpth,
            (log) => {
                if (log) {
                    log("func");
                }
                let $scope$name: Nullable<name>;
                let $scope$args: Nullable<args>;
                let $scope$body: Nullable<executable>;
                let $$res: Nullable<func> = null;
                if (true
                    && this.loop<ws>(() => this.matchws($$dpth + 1, $$cr), true) !== null
                    && this.regexAccept(String.raw`(?:public)`, $$dpth + 1, $$cr) !== null
                    && this.loop<space>(() => this.matchspace($$dpth + 1, $$cr), false) !== null
                    && this.regexAccept(String.raw`(?:function)`, $$dpth + 1, $$cr) !== null
                    && this.loop<space>(() => this.matchspace($$dpth + 1, $$cr), false) !== null
                    && ($scope$name = this.matchname($$dpth + 1, $$cr)) !== null
                    && this.loop<space>(() => this.matchspace($$dpth + 1, $$cr), true) !== null
                    && ($scope$args = this.matchargs($$dpth + 1, $$cr)) !== null
                    && this.loop<ws>(() => this.matchws($$dpth + 1, $$cr), true) !== null
                    && this.regexAccept(String.raw`(?:\{)`, $$dpth + 1, $$cr) !== null
                    && ($scope$body = this.matchexecutable($$dpth + 1, $$cr)) !== null
                    && this.regexAccept(String.raw`(?:\})`, $$dpth + 1, $$cr) !== null
                ) {
                    $$res = {kind: ASTKinds.func, name: $scope$name, args: $scope$args, body: $scope$body};
                }
                return $$res;
            }, $$cr)();
    }
    public matchargs($$dpth: number, $$cr?: ContextRecorder): Nullable<args> {
        return this.runner<args>($$dpth,
            (log) => {
                if (log) {
                    log("args");
                }
                let $scope$leadingArgs: Nullable<args_$0[]>;
                let $scope$lastArg: Nullable<Nullable<args_$1>>;
                let $$res: Nullable<args> = null;
                if (true
                    && this.regexAccept(String.raw`(?:\()`, $$dpth + 1, $$cr) !== null
                    && this.loop<ws>(() => this.matchws($$dpth + 1, $$cr), true) !== null
                    && ($scope$leadingArgs = this.loop<args_$0>(() => this.matchargs_$0($$dpth + 1, $$cr), true)) !== null
                    && this.loop<ws>(() => this.matchws($$dpth + 1, $$cr), true) !== null
                    && (($scope$lastArg = this.matchargs_$1($$dpth + 1, $$cr)) || true)
                    && this.loop<ws>(() => this.matchws($$dpth + 1, $$cr), true) !== null
                    && this.regexAccept(String.raw`(?:\))`, $$dpth + 1, $$cr) !== null
                ) {
                    $$res = {kind: ASTKinds.args, leadingArgs: $scope$leadingArgs, lastArg: $scope$lastArg};
                }
                return $$res;
            }, $$cr)();
    }
    public matchargs_$0($$dpth: number, $$cr?: ContextRecorder): Nullable<args_$0> {
        return this.runner<args_$0>($$dpth,
            (log) => {
                if (log) {
                    log("args_$0");
                }
                let $scope$name: Nullable<name>;
                let $$res: Nullable<args_$0> = null;
                if (true
                    && ($scope$name = this.matchname($$dpth + 1, $$cr)) !== null
                    && this.matchnewLineOrComma($$dpth + 1, $$cr) !== null
                    && this.loop<ws>(() => this.matchws($$dpth + 1, $$cr), true) !== null
                ) {
                    $$res = {kind: ASTKinds.args_$0, name: $scope$name};
                }
                return $$res;
            }, $$cr)();
    }
    public matchargs_$1($$dpth: number, $$cr?: ContextRecorder): Nullable<args_$1> {
        return this.matchname($$dpth + 1, $$cr);
    }
    public test(): boolean {
        const mrk = this.mark();
        const res = this.matchglobals(0);
        const ans = res !== null && this.finished();
        this.reset(mrk);
        return ans;
    }
    public parse(): ParseResult {
        const mrk = this.mark();
        const res = this.matchglobals(0);
        if (res && this.finished()) {
            return new ParseResult(res, null);
        }
        this.reset(mrk);
        const rec = new ErrorTracker();
        this.matchglobals(0, rec);
        return new ParseResult(res,
            rec.getErr() ?? new SyntaxErr(this.mark(), new Set(["$EOF"]), new Set([])));
    }
    public mark(): PosInfo {
        return this.pos;
    }
    private loop<T>(func: $$RuleType<T>, star: boolean = false): Nullable<T[]> {
        const mrk = this.mark();
        const res: T[] = [];
        for (;;) {
            const t = func();
            if (t === null) {
                break;
            }
            res.push(t);
        }
        if (star || res.length > 0) {
            return res;
        }
        this.reset(mrk);
        return null;
    }
    private runner<T>($$dpth: number, fn: $$RuleType<T>, cr?: ContextRecorder): $$RuleType<T> {
        return () => {
            const mrk = this.mark();
            const res = cr ? (() => {
                const extraInfo: string[] = [];
                const result = fn((msg: string) => extraInfo.push(msg));
                cr.record(mrk, $$dpth, result, this.negating, extraInfo);
                return result;
            })() : fn();
            if (res !== null) {
                return res;
            }
            this.reset(mrk);
            return null;
        };
    }
    private choice<T>(fns: Array<$$RuleType<T>>): Nullable<T> {
        for (const f of fns) {
            const res = f();
            if (res !== null) {
                return res;
            }
        }
        return null;
    }
    private regexAccept(match: string, dpth: number, cr?: ContextRecorder): Nullable<string> {
        return this.runner<string>(dpth,
            (log) => {
                if (log) {
                    if (this.negating) {
                        log("$$!StrMatch");
                    } else {
                        log("$$StrMatch");
                    }
                    // We substring from 3 to len - 1 to strip off the
                    // non-capture group syntax added as a WebKit workaround
                    log(match.substring(3, match.length - 1));
                }
                const reg = new RegExp(match, "y");
                reg.lastIndex = this.mark().overallPos;
                const res = reg.exec(this.input);
                if (res) {
                    let lineJmp = 0;
                    let lind = -1;
                    for (let i = 0; i < res[0].length; ++i) {
                        if (res[0][i] === "\n") {
                            ++lineJmp;
                            lind = i;
                        }
                    }
                    this.pos = {
                        overallPos: reg.lastIndex,
                        line: this.pos.line + lineJmp,
                        offset: lind === -1 ? this.pos.offset + res[0].length : (res[0].length - lind - 1)
                    };
                    return res[0];
                }
                return null;
            }, cr)();
    }
    private noConsume<T>(fn: $$RuleType<T>): Nullable<T> {
        const mrk = this.mark();
        const res = fn();
        this.reset(mrk);
        return res;
    }
    private negate<T>(fn: $$RuleType<T>): Nullable<boolean> {
        const mrk = this.mark();
        const oneg = this.negating;
        this.negating = !oneg;
        const res = fn();
        this.negating = oneg;
        this.reset(mrk);
        return res === null ? true : null;
    }
}
export function parse(s: string): ParseResult {
    const p = new Parser(s);
    return p.parse();
}
export class ParseResult {
    public ast: Nullable<globals>;
    public err: Nullable<SyntaxErr>;
    constructor(ast: Nullable<globals>, err: Nullable<SyntaxErr>) {
        this.ast = ast;
        this.err = err;
    }
}
export interface PosInfo {
    readonly overallPos: number;
    readonly line: number;
    readonly offset: number;
}
export class SyntaxErr {
    public pos: PosInfo;
    public exprules: string[];
    public expmatches: string[];
    constructor(pos: PosInfo, exprules: Set<string>, expmatches: Set<string>) {
        this.pos = pos;
        this.exprules = [...exprules];
        this.expmatches = [...expmatches];
    }
    public toString(): string {
        return `Syntax Error at line ${this.pos.line}:${this.pos.offset}. Tried to match rules ${this.exprules.join(", ")}. Expected one of ${this.expmatches.map((x) => ` '${x}'`)}`;
    }
}
class ErrorTracker implements ContextRecorder {
    private mxpos: PosInfo = {overallPos: -1, line: -1, offset: -1};
    private mnd: number = -1;
    private prules: Set<string> = new Set();
    private pmatches: Set<string> = new Set();
    public record(pos: PosInfo, depth: number, result: any, negating: boolean, extraInfo: string[]) {
        if ((result === null) === negating) {
            return;
        }
        if (pos.overallPos > this.mxpos.overallPos) {
            this.mxpos = pos;
            this.mnd = depth;
            this.pmatches.clear();
            this.prules.clear();
        } else if (pos.overallPos === this.mxpos.overallPos && depth < this.mnd) {
            this.mnd = depth;
            this.prules.clear();
        }
        if (this.mxpos.overallPos === pos.overallPos && extraInfo.length >= 2) {
            if (extraInfo[0] === "$$StrMatch") {
                this.pmatches.add(extraInfo[1]);
            }
            if (extraInfo[0] === "$$!StrMatch") {
                this.pmatches.add(`not ${extraInfo[1]}`);
            }
        }
        if (this.mxpos.overallPos === pos.overallPos && this.mnd === depth) {
            extraInfo.forEach((x) => { if (x !== "$$StrMatch" && x !== "$$!StrMatch") { this.prules.add(x); } });
        }
    }
    public getErr(): SyntaxErr | null {
        if (this.mxpos.overallPos !== -1) {
            return new SyntaxErr(this.mxpos, this.prules, this.pmatches);
        }
        return null;
    }
}