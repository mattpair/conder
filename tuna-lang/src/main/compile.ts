import { ParseResult, globals_$0_$0, ASTKinds } from "./parser";


export function compile(p: ParseResult): globals_$0_$0[] {
    if (p.err) {
        throw Error(`Failure parsing: line ${p.err.pos.line} col ${p.err.pos.offset}`)
    }
    return p.ast.map(g => {
        switch (g.value.kind) {
            case ASTKinds.varDecl: 
                if (g.value.mutability !== "const") {
                    throw Error(`Global variable ${g.value.name.name} must be const`)
                }
            case ASTKinds.func: 

                return g.value

            default: 
                const ne: never = g.value
        }
    })

}