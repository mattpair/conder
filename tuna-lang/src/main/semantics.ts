import { ParseResult, executable, ASTKinds, func, expression, literal } from "./parser";
import {AnyNode, PickNode, FunctionDescription} from "conder_core"

type GlobalObject = {kind: "glob", name: string}

export type Manifest = {
    globals: Map<string, GlobalObject>
    funcs: Map<string, FunctionDescription>
}

type ScopeMapEntry= "func" | "global" | {kind: "const" | "mut", index: number}
class ScopeMap extends Map<string, ScopeMapEntry>  {
    nextVar: number = 0
    
    public get nextVariableIndex(): number {
        return this.nextVar
    }

    set(k: string, v: ScopeMapEntry): this {
        if (v !== "global" && v !== "func") {
            this.nextVar++
        }
        super.set(k, v)
        return this
    }


}


function excluding<EX extends AnyNode["kind"]>(node: AnyNode, ...not: EX[]): Exclude<AnyNode, {kind: EX}> {
    //@ts-ignore
    if (not.includes(node.kind)) {
        throw Error(`Invalid expression ${node.kind}`)
    }
    //@ts-ignore
    return node
}
function only<IN extends AnyNode["kind"]>(node: AnyNode, ...only: IN[]): Extract<AnyNode, {kind: IN}> {
    //@ts-ignore
    if (only.includes(node.kind)) {
        //@ts-ignore
        return node
    }

    throw Error(`Invalid expression ${node.kind}`)
}

function cbOnly<IN extends AnyNode["kind"]>(node: AnyNode, f: (i: PickNode<IN>) => void, ...only: IN[]) {
    //@ts-ignore
    if (only.includes(node.kind)) {
        //@ts-ignore
        f(node)
    }
}


function literal_to_node(lit: literal, scope: ScopeMap): AnyNode {
    switch (lit.kind) {
        case ASTKinds.bool:
            return {kind: "Bool", value: lit.value === "true"}
        case ASTKinds.obj:
            
            return {
                kind: "Object", 
                fields: lit.fields.value.map(f => {
                
                    return {
                        kind: "SetField", 
                        field_name: [{kind: "String", value: f.name.name}], 
                        value: excluding(literal_to_node(f.value, scope),  "Return", "SetField", "If", "Save", "Update")
                    }
                })
            }
        case ASTKinds.str:
            return {
                kind: "String",
                value: lit.value
            }
        case ASTKinds.num:    
            return {
                kind: "Int",
                value: parseFloat(lit.value)
            }
    }
}

function method_to_node(target: PickNode<"Saved" | "GlobalObject">, methods: expression["methods"], scope: ScopeMap): AnyNode {
    if (methods.length === 0) {
        return target
    }
    const fields: PickNode<"GetField">["field_name"] = []
    methods.forEach(m => {
        switch(m.method.kind) {
            case ASTKinds.literalIndex: 
                fields.push({kind: "String", value: m.method.value.name})
        
                break
            case ASTKinds.parameterIndex:
                const e = expression_to_node(m.method.value, scope)
                fields.push(only(e, "String", "Saved"))
                break

            default: 
                const n: never = m.method
        }
    })
    
    return {
        kind: "GetField",
        target,
        field_name: fields
    }
}

type Target = {root: PickNode<"Update">["target"], field: PickNode<"SetField">["field_name"]}
function expression_to_update_target(exp: expression, scope: ScopeMap): Target {

    switch (exp.root.kind) {
        case ASTKinds.name:
            const name = scope.get(exp.root.name)
            if (name === undefined) {
                throw Error(`Unrecognized name ${exp.root.name}`)
            }
            if (name === "func") {
                throw Error(`Invalid function reference ${exp.root.name}`)
            }
            if (name !== "global" && name.kind === "const") {
                throw Error(`Attempting to overwrite constant variable ${exp.root.name}`)
            }
            
            const root: Target["root"] = name === "global" ? 
                {kind: "GlobalObject", name: exp.root.name} 
                : {kind: "Saved", index: name.index}
            
            const field: Target["field"] = exp.methods.map(m => {
                switch (m.method.kind) {
                    case ASTKinds.literalIndex:
                        return {
                            kind: "String",
                            value: m.method.value.name
                        }
                        
                    case ASTKinds.parameterIndex:
                        return only(expression_to_node(m.method.value, scope), "String", "Saved")
                    default: 
                        const n: never = m.method
              }
            })
            return {root, field}


        default: throw Error(`Invalid assignment to ${exp.root.kind}`)
    }
}

function expression_to_node(exp: expression, scope: ScopeMap): AnyNode {
    switch(exp.root.kind) {
        case ASTKinds.bool:
        case ASTKinds.str:
        case ASTKinds.num:
        case ASTKinds.obj:
            if (exp.methods.length > 0) {
                throw Error(`Unexpected method on ${exp.root.kind} literal`)
            }
            return only(literal_to_node(exp.root, scope), "Bool", "Int", "String", "Object")
            
        
        case ASTKinds.name:
            const name = scope.get(exp.root.name)
            if (name === undefined) {
                throw Error(`Unrecognized name ${exp.root.name}`)
            }
            if (name === "func") {
                throw Error(`Invalid function reference ${exp.root.name}`)
            } else if (name === "global") {
                return method_to_node({kind: "GlobalObject", name: exp.root.name}, exp.methods, scope)
            } else {
                return method_to_node({kind: "Saved", index: name.index}, exp.methods, scope)
            }
        default: 
            const n: never = exp.root
    }
    
}

function to_computation(ex: executable, scope: ScopeMap): FunctionDescription["computation"] {
    const ret: FunctionDescription["computation"] = []
    ex.value.forEach(e => {
        
        switch (e.value.kind) {
            case ASTKinds.ret:
                ret.push({
                    kind: "Return", 
                    value: e.value.value !== null ? excluding(expression_to_node(e.value.value.exp, scope), "Update", "SetField", "Return", "If", "Save") : undefined
                    }
                )
                break
            

            case ASTKinds.assignment: {
                const target: Target = expression_to_update_target(e.value.target, scope)
                const value: PickNode<"Update">["operation"] = excluding(expression_to_node(e.value.value, scope), "Update", "Return", "If", "Save", "SetField")
                
                ret.push({
                    kind: "Update",
                    target: target.root,
                    operation: target.field.length > 0 ? {kind: "SetField", field_name: target.field, value} : value
                })
                break
            }
            case ASTKinds.expression: {
                cbOnly(expression_to_node(e.value, scope), ret.push, "Return", "If", "Save", "Update")
                
                break
            }

            case ASTKinds.varDecl: 
                const value = excluding(expression_to_node(e.value.value, scope), "Return", "If", "Update", "Save", "SetField")
                const index = scope.nextVariableIndex
                if (scope.has(e.value.name.name)) {
                    throw Error(`The symbol ${e.value.name.name} is already in use`)
                }
                scope.set(e.value.name.name, {kind: e.value.mutability === "const" ? "const" : "mut", index})
                ret.push({
                    kind: "Save",
                    value,
                    index
                })
                break
            default: 
                const n: never = e.value
        }
    })
    return ret
}

function to_descr(f: func, scope: ScopeMap): FunctionDescription {
    try {
        const argList: string[] = []
        if (f.args.leadingArgs.length > 0) {
            argList.push(...f.args.leadingArgs.map(a => a.name.name))
        }
        if (f.args.lastArg) {
            argList.push(f.args.lastArg.name)
        }
        const input: FunctionDescription["input"] = []
        argList.forEach((a, i) => {
            if (scope.has(a)) {
                throw Error(`Arg ${a} is using a name already in use`)
            }
            scope.set(a, {kind: "mut", index: i})
            input.push({kind: "Any", data: undefined})
        })
        
        return {
            input,
            computation: to_computation(f.body, scope)
        }
    } catch(e) {
        throw Error(`In function ${f.name.name}: \n\t${e.message}`)
    }
    
}

export function semantify(p: ParseResult): Manifest {
    if (p.err) {
        throw Error(`Failure parsing: line ${p.err.pos.line} col ${p.err.pos.offset}: ${p.err.toString()}`)
    }

    const globalScope = new ScopeMap()
    const aFunc: func[] = []

    const funcs: Map<string, FunctionDescription> = new Map()
    const globs: Map<string, GlobalObject> = new Map()

    p.ast.forEach(g => {
        const name = g.value.name.name
        if (globalScope.has(name)) {
            throw Error(`Another global or function is already called ${name}`)
        }
        
        switch (g.value.kind) {
        
            case ASTKinds.varDecl: 
                if (g.value.mutability !== "const") {
                    throw Error(`Global variable ${name} must be const`)
                }
                
                switch (g.value.value.root.kind) {
                    case ASTKinds.obj: 
                        if (g.value.value.root.fields.value.length > 0) {
                            break
                        }
                        if (g.value.value.methods.length > 0) {
                            break
                        }
                        globalScope.set(name, "global")
                        globs.set(name, {kind: "glob", name})
                        return
                    default: 
                        break
                }

                throw Error(`Global ${name} must be initialized as empty object`)
                
            case ASTKinds.func: 
                globalScope.set(name, "func")
                aFunc.push(g.value)
                break
            default: 
                const ne: never = g.value
        }
    })


    aFunc.forEach(f => funcs.set(f.name.name, to_descr(f, globalScope)))

    return  {
        globals: globs,
        funcs
    }


}