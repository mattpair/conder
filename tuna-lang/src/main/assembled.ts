import {FUNCTION_TRANSFORMER, FunctionDescription, Transformer ,Compiler} from 'conder_core'
import {Parser} from './parser'
import {semantify, Manifest} from './semantics'

export const TUNA_TO_MANIFEST = new Transformer<string, Manifest>(str => {
    const p = new Parser(str).parse()
    return semantify(p)
})