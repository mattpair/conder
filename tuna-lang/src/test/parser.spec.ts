import { Parser } from '../main/parser';


describe("parser", () => {
    it("should parse a global object", () => {
        const p = new Parser(`const obj = {}`)
        const r = p.parse()
        expect(r).toMatchSnapshot()
    })
})