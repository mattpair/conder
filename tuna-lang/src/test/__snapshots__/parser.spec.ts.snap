// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parser should parse a global object 1`] = `
ParseResult {
  "ast": Object {
    "kind": "initializeConst",
    "name": Object {
      "kind": "name",
      "name": "obj",
    },
    "value": Object {
      "kind": "obj",
    },
  },
  "err": null,
}
`;
